### PR TITLE
Make RocksDB secondary instance respect atomic groups in version edits

### PR DESCRIPTION
**Pull request by HaoyuHuang**
With this commit, RocksDB secondary instance respects atomic groups in version edits.

### CODE REVIEW

File: db/version_set.cc [hunk 1]
```diff
@@ -5457,52 +5481,109 @@ Status ReactiveVersionSet::ReadAndApply(
...
+    bool* have_log_number, uint64_t* log_number, bool* have_prev_log_number,
+    uint64_t* previous_log_number, bool* have_next_file, uint64_t* next_file,
+    bool* have_last_sequence, SequenceNumber* last_sequence,
+    uint64_t* min_log_number_to_keep, uint32_t* max_column_family) {
+  ColumnFamilyData* cfd =
+      column_family_set_->GetColumnFamily(edit.column_family_);
+
+  // If we cannot find this column family in our column family set, then it
+  // may be a new column family created by the primary after the secondary
+  // starts. Ignore it for now.
```
Comment (riversand963) [inline db/version_set.cc]: It is also possible that the secondary instance opens only a subset of column families. Maybe add a sentence or two here to account for this case as well.

File: db/version_set.h [hunk 2]
```diff
@@ -747,6 +747,23 @@ struct ObsoleteFileInfo {
 
 class BaseReferencedVersionBuilder;
 
+class AtomicGroupReadBuffer {
+ public:
+  Status AddEdit(VersionEdit* edit);
+  void Clear();
+  bool IsFull();
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.h]: Thanks! Added const.

File: db/version_set.cc [hunk 3]
```diff
@@ -4071,6 +4116,74 @@ Status VersionSet::GetCurrentManifestPath(const std::string& dbname, Env* env,
...
+      if (s.ok()) {
+        recovered_edits++;
+      }
+    }
+  }
+  if (!s.ok()) {
+    // Clear the buffer if we fail to decode/apply an edit.
+    read_buffer->Clear();
+  }
+  TEST_SYNC_POINT_CALLBACK("VersionSet::Recover:RecoveredEdits",
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.cc]: My mistake. I updated the name. This sync point does not exist before. I added it for testing.

File: db/version_edit.h [hunk 4]
```diff
@@ -316,6 +316,7 @@ class VersionEdit {
   friend class ReactiveVersionSet;
   friend class VersionSet;
   friend class Version;
+  friend class AtomicGroupReadBuffer;
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_edit.h]: I did run make format. But it does not show any error for this. What's the problem with this line?

File: db/version_set.h [hunk 5]
```diff
@@ -1123,16 +1152,23 @@ class ReactiveVersionSet : public VersionSet {
                  std::unique_ptr<log::Reader::Reporter>* manifest_reporter,
                  std::unique_ptr<Status>* manifest_reader_status);
 
+  uint64_t read_edits_in_atomic_group() const {
+    return read_buffer.read_edits_in_atomic_group();
+  }
+  std::vector<VersionEdit> replay_buffer() const {
+    return read_buffer.replay_buffer();
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.h]: No. It is not. I changed it to use reference.

File: db/version_set.h [hunk 6]
```diff
@@ -1123,16 +1152,23 @@ class ReactiveVersionSet : public VersionSet {
                  std::unique_ptr<log::Reader::Reporter>* manifest_reporter,
                  std::unique_ptr<Status>* manifest_reader_status);
 
+  uint64_t read_edits_in_atomic_group() const {
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.h]: Yes. I updated the function name.

File: db/version_set.h [hunk 7]
```diff
@@ -1141,6 +1177,7 @@ class ReactiveVersionSet : public VersionSet {
  private:
   std::unordered_map<uint32_t, std::unique_ptr<BaseReferencedVersionBuilder>>
       active_version_builders_;
+  AtomicGroupReadBuffer read_buffer;
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.h]: Updated.

File: db/version_set.cc (hunk 1 above)
Comment (HaoyuHuang) [reply db/version_set.cc]: I added the comments inline.

File: db/version_set.cc [hunk 8]
```diff
@@ -5350,7 +5401,7 @@ Status ReactiveVersionSet::ReadAndApply(
   uint64_t previous_log_number = 0;
   uint32_t max_column_family = 0;
   uint64_t min_log_number_to_keep = 0;
-
+  uint64_t applied_edits = 0;
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/version_set.cc]: I added this variable for testing ReadAndApply.

File: db/version_set.cc [hunk 9]
```diff
@@ -5457,52 +5481,110 @@ Status ReactiveVersionSet::ReadAndApply(
...
+  }
+
+  if (cfd != nullptr) {
+    s = builder->LoadTableHandlers(
+        cfd->internal_stats(), db_options_->max_file_opening_threads,
+        false /* prefetch_index_and_filter_in_cache */,
+        false /* is_initial_load */,
+        cfd->GetLatestMutableCFOptions()->prefix_extractor.get());
+    TEST_SYNC_POINT_CALLBACK(
+        "ReactiveVersionSet::ReadAndApply:AfterLoadTableHandlers", &s);
```
Comment (riversand963) [inline db/version_set.cc]: Sync point naming?

File: db/version_set_test.cc [hunk 10]
```diff
@@ -758,12 +763,29 @@ TEST_F(VersionSetTest, SameColumnFamilyGroupCommit) {
...
+ protected:
+  std::unique_ptr<log::Writer> log_writer_;
+};
+
+const int VersionSetAtomicGroupTest::kTestVersionSetRecover = 0;
+const int VersionSetAtomicGroupTest::kTestReactiveVersionSetRecover = 1;
+const int VersionSetAtomicGroupTest::kTestReactiveVersionSetReadAndApply = 2;
+
+TEST_P(VersionSetAtomicGroupTest, HandleValidAtomicGroup) {
+  const int test_use_case = GetParam();
```
Comment (riversand963) [inline db/version_set_test.cc]: If you use GetParam() to distinguish different test cases, why not make them separate tests?

File: db/version_set.cc (hunk 9 above)
Comment (HaoyuHuang) [reply db/version_set.cc]: Renamed. Thanks!

File: db/version_set_test.cc (hunk 10 above)
Comment (HaoyuHuang) [reply db/version_set_test.cc]: I separate them into different tests.

**Comment by HaoyuHuang** [comment]
I rebased my local branch so that the appveyor and travis checks would run.

**Comment by HaoyuHuang** [comment]
My local branch is too old and causes this PR to show 3000 lines of code changes after I rebase. I'm going to send out a new PR

**Pull request closed without merging**